File: subprojects/basic03/dynamic_src.py
#!/usr/bin/env python3

"""
Supplement to Basic Tutorial 3: Dynamic Source

Simple example to demonstrate dynamically adding and removing source elements
to a playing pipeline.
"""

import sys
import logging

import gi
gi.require_version("Gst", "1.0")
gi.require_version("GLib", "2.0")
from gi.repository import GLib, Gst

logging.basicConfig(level=logging.INFO, format="[%(levelname)8s] - %(message)s")
logger = logging.getLogger(__name__)


class ProbeData:
    def __init__(self, pipeline, source, sink, index):
        self.pipeline = pipeline
        self.source = source
        self.sink = sink
        self.index = index


def on_bus_message(bus, message, loop):
    if message.type == Gst.MessageType.EOS:
        logger.info("End-of-stream reached.")
        loop.quit()
    elif message.type == Gst.MessageType.ERROR:
        err, debug_info = message.parse_error()
        logger.error(f"Error received from element {message.src.get_name()}: {err.message}")
        logger.error(f"Debugging information: {debug_info if debug_info else 'none'}")
        loop.quit()
    elif message.type == Gst.MessageType.STATE_CHANGED:
        # We are only interested in state-changed messages from the pipeline
        old_state, new_state, pending_state = message.parse_state_changed()
        if ("test-pipeline" == message.src.get_name()):
            logger.info(f"Pipeline state changed: {old_state.value_nick} -> {new_state.value_nick}")
    else:
        logger.debug(f"Unhandled message type {message.type}.")
    return True


def on_timeout(pdata):
    pdata.source.set_state(Gst.State.NULL)
    pdata.pipeline.remove(pdata.source)
    # Create a new source
    pdata.source = Gst.ElementFactory.make("videotestsrc", "source")
    pdata.index = pdata.index + 1 if pdata.index < 25 else 0
    pdata.source.props.pattern = pdata.index
    pdata.source.props.is_live = True
    # rebuild the pipeline
    pdata.pipeline.add(pdata.source)
    pdata.source.link(pdata.sink)
    pdata.source.set_state(Gst.State.PLAYING)
    # start the next timeout
    GLib.timeout_add_seconds(1, on_timeout, pdata)


def main(argv):
    # Initialize GStreamer
    Gst.init(argv[1:])

    # Create the elements
    source = Gst.ElementFactory.make("videotestsrc", "source")
    sink = Gst.ElementFactory.make("autovideosink", "sink")

    # Create the empty pipeline
    pipeline = Gst.Pipeline.new("test-pipeline")

    if not pipeline or not source or not sink:
        logger.error("Not all elements could be created.")
        return -1

    # Build the pipeline.
    pipeline.add(source, sink)
    # Link the source and sink
    if not source.link(sink):
        logger.error("Elements could not be linked.")
        return -2

    # Set the pattern property of the "videotestsrc"
    source.set_property("pattern", 0)

    loop = GLib.MainLoop()

    # Listen to the bus
    bus = pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect("message", on_bus_message, loop)

    # Prepare data for callback
    pdata = ProbeData(pipeline, source, sink, 0)

    GLib.timeout_add_seconds(1, on_timeout, pdata)

    # start play back and listen to events
    ret = pipeline.set_state(Gst.State.PLAYING)
    if ret == Gst.StateChangeReturn.FAILURE:
        logger.error("Unable to set the pipeline to the playing state.")
        return -1

    try:
        loop.run()
    except:
        logger.error("Unhandled exception.")

    # cleanup
    pipeline.set_state(Gst.State.NULL)


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: subprojects/basic03/dynamic_src.cpp
/* gstreamermm - a C++ wrapper for gstreamer
 *
 * Supplement to Basic Tutorial 3: Dynamic Source
 *
 * Simple example to demonstrate dynamically adding and removing source elements
 * to a playing pipeline.
 */

#include <gstreamermm.h>
#include <glibmm/main.h>
#include <iostream>
#include <cstdlib>

using Glib::RefPtr;

// Expose objects to be used in callbacks
Glib::RefPtr<Glib::MainLoop> mainloop;
RefPtr<Gst::Element> source;
RefPtr<Gst::Element> sink;
RefPtr<Gst::Pipeline> pipeline;

// This function is used to receive asynchronous messages in the main loop.
bool on_bus_message(const RefPtr<Gst::Bus>&,
    const RefPtr<Gst::Message>& message)
{
  switch(message->get_message_type())
  {
    case Gst::MESSAGE_EOS:
      std::cout << std::endl << "End of stream" << std::endl;
      mainloop->quit();
      return false;
    case Gst::MESSAGE_ERROR:
    {
      RefPtr<Gst::MessageError> msgError {RefPtr<Gst::MessageError>::cast_static(message)};
      if (msgError)
      {
        Glib::Error err {msgError->parse_error()};
        std::string debug_info {msgError->parse_debug()};
        std::cerr << "Error received from element " << message->get_source()->get_name() << ": " <<
            err.what() << std::endl;
        if (!debug_info.empty())
        {
          std::cout << "Debugging information: " << debug_info << std::endl;
        }
      }
      else
      {
        std::cerr << "Error." << std::endl;
      }
      mainloop->quit();
      return false;
    }
    case Gst::MESSAGE_STATE_CHANGED:
    {
      // We are only interested in state-changed messages from the pipeline
      if ("test-pipeline" == message->get_source()->get_name())
      {
        auto state_get_name = [] (Gst::State state) -> std::string {
          return gst_element_state_get_name(static_cast<GstState>(state));
        };
        RefPtr<Gst::MessageStateChanged> msgSC {RefPtr<Gst::MessageStateChanged>::cast_static(message)};
        Gst::State old_state {msgSC->parse_old_state()};
        Gst::State new_state {msgSC->parse_new_state()};
        std::cout << "Pipeline state changed: " <<
            state_get_name(old_state) << " -> " <<
            state_get_name(new_state) << std::endl;
      }
      break;
    }
    default:
        //std::cout << "Unhandled message type: " << message->get_message_type() << std::endl;
      break;
  }

  return true;
}

bool on_timeout()
{
	static int pattern = 0;

	source->set_state(Gst::STATE_NULL);
  pipeline->remove(source);
  // Create a new source
  source = Gst::ElementFactory::create_element("videotestsrc", "source");
  pattern = (pattern < 25) ? (pattern + 1) : 0;
  source->set_property("pattern", pattern);
  source->set_property("is_live", true);
  // Rebuild the pipeline
  pipeline->add(source);
  source->link(sink);
	source->set_state(Gst::STATE_PLAYING);

	return true;
}

int main(int argc, char** argv)
{
  // Initialize gstreamermm:
  Gst::init(argc, argv);

  // Create elements
  source = Gst::ElementFactory::create_element("videotestsrc", "source");
  sink = Gst::ElementFactory::create_element("autovideosink", "sink");

  // Create the empty pipeline
  pipeline = Gst::Pipeline::create("test-pipeline");

  if (!source || !sink || !pipeline)
  {
    std::cerr << "Pipeline or one of the elements could not be created." << std::endl;
    return EXIT_FAILURE;
  }

  try
  {
    // add the elements to the pipeline before linking them
    pipeline->add(source)->add(sink);
    // Link the source and sink
    source->link(sink);
  }
	catch (const std::exception& ex)
  {
		std::cerr << "Exception occured during preparing pipeline: " << ex.what() << std::endl;
    return EXIT_FAILURE;
  }

  // Set the URI to play
  source->set_property("pattern", 0);

  // Create the main loop.
  mainloop = Glib::MainLoop::create();

  // Get the bus and watch the messages
  RefPtr<Gst::Bus> bus {pipeline->get_bus()};
  bus->add_watch(sigc::ptr_fun(&on_bus_message));

  // start play back and listen to events
  if (pipeline->set_state(Gst::STATE_PLAYING) == Gst::STATE_CHANGE_FAILURE)
  {
    std::cerr << "Unable to set the pipeline to the playing state." << std::endl;
    return EXIT_FAILURE;
  }

	Glib::signal_timeout().connect(sigc::ptr_fun(&on_timeout), 1000);

  // Now set the playbin to the PLAYING state and start the main loop:
  std::cout << "Running." << std::endl;
  mainloop->run();

  // Clean up nicely:
  std::cout << "Returned. Stopping pipeline." << std::endl;
  pipeline->set_state(Gst::STATE_NULL);

  return EXIT_SUCCESS;
}
